src/TimeSeries/Backend/Graphite/GraphiteBackend.php

```diff
@@ -0,0 +1,188 @@
+    const QUERY_TIMEOUT = 120;
+    const DATA_CACHE_TIMEOUTS = [
+        7200 => 60, // cache last two hours for 1 min
+        86400 => 600, // cache last day for 10 min
+    ];
+    const DATA_CACHE_TIMEOUT_DEFAULT = 3600;
+
+    /**
+     * @var ExpressionFactory
+     */
+    private $expressionFactory;
+
+    public function __construct(ExpressionFactory $expressionFactory){
+        $this->expressionFactory = $expressionFactory;
+    }
+
+    /**
+     * Make a query to the graphite backend service.
+     *
+     * @param string $path
+     * @param array $params
+     * @return string
+     * @throws BackendException
+     */
+    private function sendQuery(string $path, array $params): string
+    {
+        $query = http_build_query($params);
+        // hax to replace target[0]...target[1]... with []
+        $query = preg_replace('/%5B[0-9]+%5D/simU', '%5B%5D', $query);
+
+        $url = GraphiteBackend::GRAPHITE_URL . $path;
+        $ch = curl_init($url);
+        curl_setopt_array($ch,
+                          array(
+                              CURLOPT_TIMEOUT => GraphiteBackend::QUERY_TIMEOUT,
+                              CURLOPT_RETURNTRANSFER => true,
+                              CURLOPT_FAILONERROR => true,
+                              CURLOPT_POSTFIELDS => $query,
+                          )
+        );
+        $result = curl_exec($ch);
+        if ($result === false || !$result) {
+            throw new BackendException(curl_error($ch) . " URL: " . $url . '?' . $query);
+        }
+        curl_close($ch);
+        return $result;
+    }
+
+    private function calcTimeout($until){
+        // if until is a relative time, then we want a low cache timeout
+        $now = time();
+        $timeout = GraphiteBackend::DATA_CACHE_TIMEOUT_DEFAULT;
+        if ($until->isRelative()) {
+            $t = $now;
+        } else {
+            $t = $until->getAbsoluteTime()->getTimestamp();
+        }
+        foreach (GraphiteBackend::DATA_CACHE_TIMEOUTS as $limit => $tout) {
+            if ($t >= $now - $limit) {
+                // our 'until' time falls inside this limit
+                $timeout = $tout;
+                break;
+            }
+        }
+
+        return $timeout;
+    }
+
+    /**
+     * Construct graphite query, send request, and return one TimeSeries object.
+     *
+     * @param QueryTime $from
+     * @param QueryTime $until
+     * @param $expressionJson
+     * @param int|null $maxPoints
+     * @return TimeSeries
+     * @throws BackendException|Expression\ParsingException
+     */
+    public function queryGraphite(QueryTime $from, QueryTime $until, $expressionJson, ?int $maxPoints): TimeSeries
+    {
+        // calculate cache timeout, used in graphite query
+        $timeout = $this->calcTimeout($until);
+
+        // build expressions from JSON
+        $exp = $this->expressionFactory->createFromJson($expressionJson) -> getCanonicalStr();
```

Comment:
oh wow, yeah i think you really want to change to just directly passing in the graphite query. you're now building the expression object for a second time, just to throw it away and get the graphite query string that you could have passed in the first place.